bool change
int32 value
